### PR TITLE
Cluster and Datastore selection when hosts are under maintenance.

### PR DESCRIPTION
**Pull request by EleanorRigby**
Add support for host under maintenance mode
Add unit test for host in maintenance mode changes.
Refactor picker code and datastore(Resource) class
Add == operator to managed object class.

Passes all unit and integrations test.

### CODE REVIEW

File: src/vsphere_cpi/lib/cloud/vsphere/cluster_picker.rb [hunk 1]
```diff
@@ -32,6 +32,21 @@ def best_cluster_placement(req_memory:, disk_configurations:)
...
+      # Filter out placements where the
+      # datastore has no active hosts i.e. (hosts are under maintenance) or
+      # if there are any active host, they belong to a different cluster
+      placement_options = placements_with_active_hosts(placement_options, clusters)
+      if placement_options.size == 0
+        raise Bosh::Clouds::CloudError,
+              "No valid placement found due to no active host (non-maintenance) present on desired cluster"
+      end
+
       if placement_options.size == 0
```
Comment (nehagjain15) [inline src/vsphere_cpi/lib/cloud/vsphere/cluster_picker.rb]: Remove this redundant if block

File: src/vsphere_cpi/lib/cloud/vsphere/datastore_picker.rb [hunk 2]
```diff
@@ -50,25 +50,55 @@ def best_disk_placement(disks)
...
+      raise Bosh::Clouds::CloudError, pretty_print_placement_error(disks) if placement[:datastores].empty?
+
       add_balance_score(placement)
+
+    end
+
+    def filter_all_placement_without_disks(placement)
+      datastore_placements = placement[:datastores]
+
+      iterate_placement = datastore_placements.clone
```
Comment (nehagjain15) [inline src/vsphere_cpi/lib/cloud/vsphere/datastore_picker.rb]: you can iterate over same hash as well as update it at the same time, so you don't need to clone

File: src/vsphere_cpi/spec/integration/host_maintenance_mode_spec.rb [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+          ]
+      )
+      VSphereCloud::Cloud.new(options)
+    end
+
+    # We turn the maintenance mode ON for all of the hosts (rounding off to the floor) in a cluster
+    before do
+      turn_maintenance_on_for_all_hosts(cpi, @cluster_name_maintenance)
+    end
+    it 'cpi should fail to replicate stemcell (create vm and create ephemeral disk) on a datastore accessible by all maintenance mode hosts and which belongs to same cluster' do
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/integration/host_maintenance_mode_spec.rb]: This it statement can be just 'cpi should fail to replicate stemcell (create vm and create ephemeral disk)' as rest of it is covered by context.
Similarly you can update it statements for other tests and remove duplication

File: src/vsphere_cpi/spec/unit/cloud/vsphere/resources/datastore_spec.rb [hunk 4]
```diff
@@ -1,60 +1,125 @@
...
 describe VSphereCloud::Resources::Datastore do
-  subject(:datastore) {
-    VSphereCloud::Resources::Datastore.new('foo_lun', datastore_mob,  true, 16 * 1024, 8 * 1024)
-  }
-  subject(:datastore_inaccess) {
-    VSphereCloud::Resources::Datastore.new('foo_lun', datastore_mob,  false, 16 * 1024, 8 * 1024)
-  }
+  subject do
+    described_class.new('foo_lun', mob, true, 16 * 1024, 8 * 1024)
+  end
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/resources/datastore_spec.rb]: please remove unnecessary blank lines and you can apply rubymine indentation settings per https://www.pivotaltracker.com/n/projects/1133984/stories/134478283.

File: src/vsphere_cpi/spec/unit/cloud/vsphere/vm_config_spec.rb [hunk 5]
```diff
@@ -216,15 +222,23 @@ module VSphereCloud
           }
         end
         let(:global_clusters) { [cluster_1, cluster_2] }
-        let(:smaller_ds) { instance_double(VSphereCloud::Resources::Datastore, free_space: 512) }
-        let(:larger_ds) { instance_double(VSphereCloud::Resources::Datastore, free_space: 2048) }
+        let(:host_runtime_info) { instance_double(VimSdk::Vim::Host::RuntimeInfo, in_maintenance_mode: false) }
+        let(:host_system) {instance_double(VimSdk::Vim::HostSystem, runtime: host_runtime_info)}
+        let(:datastore_host_mount) { [instance_double('VimSdk::Vim::Datastore::HostMount', key: host_system)]}
+        let(:smaller_datastore_mob) { instance_double('VimSdk::Vim::Datastore', host: datastore_host_mount) }
+        let(:larger_datastore_mob) { instance_double('VimSdk::Vim::Datastore', host: datastore_host_mount) }
```
Comment (nehagjain15) [inline src/vsphere_cpi/spec/unit/cloud/vsphere/vm_config_spec.rb]: we can have a chore to see how this setup can be simplified for tests

File: src/vsphere_cpi/spec/unit/cloud/vsphere/resources/datastore_spec.rb (hunk 4 above)
Comment (EleanorRigby) [reply src/vsphere_cpi/spec/unit/cloud/vsphere/resources/datastore_spec.rb]: I did not find any blank spaces or line in this file. Did you intend to mention some other files?

**Comment by EleanorRigby** [comment]
Creating a new one to remove codacy.

**Pull request closed without merging**